File: lec13/data_race.py
import threading

N_THREADS = 50
N_INCREMENTS = 100000

my_number = 0

def thread():
    global my_number
    for _ in range(N_INCREMENTS):
        my_number += 1

threads = []

for _ in range(N_THREADS):
    my_thread = threading.Thread(target=thread)
    my_thread.start()
    threads.append(my_thread)

for thread in threads:
    thread.join()

print(f'Final total: {my_number} (expected {N_THREADS * N_INCREMENTS})')

File: lec13/DataRace.java
class DataRace {
    static final int N_THREADS = 50;
    static final int N_INCREMENTS = 100000;

    static int my_number = 0;

    static void thread() {
        for (int i = 0; i < N_INCREMENTS; i++) {
            my_number += 1;
        }
    }

    public static void main(String[] args) throws InterruptedException {
        Thread[] threads = new Thread[N_THREADS];
        
        for (int i = 0; i < N_THREADS; i++) {
            threads[i] = new Thread(DataRace::thread);
            threads[i].start();
        }
        for (int i = 0; i < N_THREADS; i++) {
            threads[i].join();
        }
        
        System.out.println(String.format("Final total: %d (expected %d)\n", my_number, N_THREADS * N_INCREMENTS));
    }

}
